examples/rand-extension/lib.rs

```diff
@@ -0,0 +1,132 @@
+            _ => panic!("encountered unknown status code"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum CustomEnvironment {}
+
+impl Environment for CustomEnvironment {
+    const MAX_EVENT_TOPICS: usize =
+        <ink_env::DefaultEnvironment as Environment>::MAX_EVENT_TOPICS;
+
+    type AccountId = <ink_env::DefaultEnvironment as Environment>::AccountId;
+    type Balance = <ink_env::DefaultEnvironment as Environment>::Balance;
+    type Hash = <ink_env::DefaultEnvironment as Environment>::Hash;
+    type BlockNumber = <ink_env::DefaultEnvironment as Environment>::BlockNumber;
+    type Timestamp = <ink_env::DefaultEnvironment as Environment>::Timestamp;
+
+    type ChainExtension = FetchRandom;
+}
+
+#[ink::contract(env = crate::CustomEnvironment)]
+mod rand_extension {
+    use super::RandomReadErr;
+
+    /// Defines the storage of your contract.
+    /// Here we store the random seed fetched from the chain
+    #[ink(storage)]
+    pub struct RandExtension {
+        /// Stores a single `bool` value on the storage.
+        value: [u8; 32],
+    }
+
+    #[ink(event)]
+    pub struct RandomUpdated {
+        #[ink(topic)]
+        new: [u8; 32],
+    }
+
+    impl RandExtension {
+        /// Constructor that initializes the `bool` value to the given `init_value`.
+        #[ink(constructor)]
+        pub fn new(init_value: [u8; 32]) -> Self {
+            Self { value: init_value }
+        }
+
+        /// Constructor that initializes the `bool` value to `false`.
+        ///
+        /// Constructors can delegate to other constructors.
+        #[ink(constructor)]
+        pub fn default() -> Self {
+            Self::new(Default::default())
+        }
+
+        /// update the value from runtime random source
+        #[ink(message)]
+        pub fn update(&mut self) -> Result<(), RandomReadErr> {
+            // Get the on-chain random seed
+            let new_random = self.env().extension().fetch_random()?;
+            self.value = new_random;
+            // emit the RandomUpdated event when the random seed
+            // is successfully fetched.
+            self.env().emit_event(RandomUpdated { new: new_random });
+            Ok(())
+        }
+
+        /// Simply returns the current value.
+        #[ink(message)]
+        pub fn get(&self) -> [u8; 32] {
+            self.value
+        }
+    }
+
+    /// Unit tests in Rust are normally defined within such a `#[cfg(test)]`
+    #[cfg(test)]
+    mod tests {
+        /// Imports all the definitions from the outer scope so we can use them here.
+        use super::*;
+        use ink_lang as ink;
+
+        /// We test if the default constructor does its job.
+        #[ink::test]
+        fn default_works() {
+            let rand_extension = RandExtension::default();
+            assert_eq!(rand_extension.get(), [0; 32]);
+        }
+    }
```

Comment:
We also provide you with a way to register chain extensions to the off-chain environment for testing purposes. The random function could probably be registered there to return a random sequence and return it as `[u8; 32]` back to the caller.
https://paritytech.github.io/ink/ink_env/test/fn.register_chain_extension.html
It might be that the API does not yet work correctly.

Reply:
@Robbepop I'll check it.

Reply:
I will merge this as it seems to be in good shape and create a follow-up issue to add an off-chain test and assign you to it, okay?

Reply:
https://github.com/paritytech/ink/issues/664